jac/jaclang/runtimelib/machine.py

```diff
@@ -469,13 +469,26 @@ def spawn_call(
         # walker ability with any exit
         for i in warch._jac_exit_funcs_:
             if not i.trigger:
-                i.func(warch, current_loc)
+                JacWalker.execute_ability(i.func, (warch, current_loc))
             if walker.disengaged:
                 return warch
 
         walker.ignores = []
         return warch
 
+    @staticmethod
+    def execute_ability(func: Callable, args: tuple) -> None:
+        """Execute abilities."""
+        if inspect.iscoroutinefunction(func):
+            try:
+                asyncio.get_running_loop()
+            except RuntimeError:
+                asyncio.run(func(*args))
+            else:
+                asyncio.ensure_future(func(*args))
+        else:
+            func(*args)
```

Comment:
Why is this type of per call dynamic detection needed? I'm concerned about performance impact of calling this for every ability call. Is this to support the dynamism alexie was asking for?